plugins/anomaly-detection/AnomalyDetection.py

```diff
@@ -0,0 +1,615 @@
+        ]))),
+        StructField('algoType', StringType(), True),
+        StructField('algoCalc', ArrayType(DoubleType(), True))
+    ])
+
+    plotDF4 = spark.createDataFrame(rdd2, rdd2Schema)
+    plotDF4 = plotDF4.withColumn("Throughputs", f.col(
+        "Diff_Secs, Throughput.max(throughput)").cast(
+        ArrayType(DecimalType(38, 18)))).drop("Diff_Secs, Throughput")
+
+    # DF to RDD to calculate anomaly using EWMA, ARIMA and DBSCAN
+    rdd3 = plotDF4.rdd.map(lambda x: (x[0], x[1], x[2], x[3], x[4], x[5],
+                                      x[6], x[7], x[8], x[9], x[10],
+                                      anomaly_func(x)))
+
+    # Schema for the Dataframe to ve created from the RDD
+    rdd3Schema = StructType([
+        StructField('sourceIP', StringType(), True),
+        StructField('sourceTransportPort', LongType(), True),
+        StructField('destinationIP', StringType(), True),
+        StructField('destinationTransportPort', LongType(), True),
+        StructField('protocolIdentifier', LongType(), True),
+        StructField('flowStartSeconds', TimestampType(), True),
+        StructField('flowEndSeconds', ArrayType(TimestampType(), True)),
+        StructField('throughputStandardDeviation', DoubleType(), True),
+        StructField('algoType', StringType(), True),
+        StructField('algoCalc', ArrayType(DoubleType(), True)),
+        StructField('Throughputs', ArrayType(DecimalType(38, 18), True)),
+        StructField('anomaly', ArrayType(BooleanType(), True))
+    ])
+
+    plotDF5 = spark.createDataFrame(rdd3, rdd3Schema)
+    ret_plotDF = filter_df_with_true_anomalies(spark, plotDF5, algo_type)
+    # Write anomalous records to DB/CSV - Module WIP
+    # Module to write to CSV. Optional.
+    ret_plotDF = ret_plotDF.withColumn(
+        "flowEndSeconds",
+        f.col("flowEndSeconds").cast("string")).withColumn(
+        "throughputStandardDeviation",
+        f.col("throughputStandardDeviation").cast("string")).withColumn(
+        "algoCalc", f.col("algoCalc").cast("string")).withColumn(
+        "Throughputs", f.col("Throughputs").cast("string")).withColumn(
+        "anomaly", f.col("anomaly").cast(
+            "string"))
+    ret_plotDF = ret_plotDF.withColumn('id', f.lit(str(tad_id_input)))
+    return ret_plotDF
+
+
+def anomaly_detection(algo_type, db_jdbc_address, start_time, end_time,
+                      tad_id_input):
+    spark = SparkSession.builder.getOrCreate()
+    sql_query = ("select {} from {} "
+                 .format(", ".join(FLOW_TABLE_COLUMNS), table_name))
+    if start_time:
+        sql_query += "WHERE flowStartSeconds >= '{}' ".format(start_time)
+    if end_time:
+        if start_time:
+            sql_query += "AND flowEndSeconds < '{}' ".format(end_time)
+        else:
+            sql_query += "WHERE flowEndSeconds < '{}' ".format(end_time)
+    sql_query += "group by {} ".format(", ".join(GROUP_COLUMNS))
+
+    plotDF = (
```

Comment:
Could you rename these DFs please? `plotDF` and `plotDF1, 2, 3...` are less meaningful IMO.

Reply:
ack